Sources/SourceKitLSP/Swift/CodeActions/ConvertJSONToCodableStruct.swift

```diff
@@ -0,0 +1,318 @@
+      text = literalText
+    }
+
+    // Try to process this as JSON.
+    guard
+      let object = try? JSONSerialization.jsonObject(with: text.data(using: .utf8)!),
+      let serial = object as? Dictionary<String, Any>
+    else {
+      return []
+    }
+
+    // Build the declarations from the JSON dictionary.
+    let decls = build(from: serial)
+
+    // Render the change into
+    switch preflight {
+    case .closure(let closure):
+      return [
+        SourceEdit(range: closure.trimmedRange, replacement: decls.description)
+      ]
+    case .tail(let closure, let unexpected):
+      return [
+        SourceEdit(range: closure.trimmedRange, replacement: decls.description),
+        SourceEdit(range: unexpected.range, replacement: ""),
+      ]
+    case .stringLiteral(let literal, _):
+      return [
+        SourceEdit(
+          range: literal.endPosition..<literal.endPosition,
+          replacement: "\n" + decls.description
+        )
+      ]
+    }
+  }
+}
+
+extension ConvertJSONToCodableStruct {
+  /// The result of preflighting a syntax node to try to find potential JSON
+  /// in it.
+  public enum Preflight {
+    /// A closure, which is what a JSON dictionary looks like when pasted
+    /// into Swift.
+    case closure(ClosureExprSyntax)
+
+    /// A closure at the end of a source file, which might have extra JSON.
+    case tail(ClosureExprSyntax, UnexpectedNodesSyntax)
+
+    /// A string literal that may contain JSON.
+    case stringLiteral(StringLiteralExprSyntax, String)
+  }
+
+  /// Look for either a closure or a string literal that might have JSON in it.
+  public static func preflightRefactoring(_ syntax: Syntax) -> Preflight? {
+    // Preflight a closure.
+    //
+    // A blob of JSON dropped into a Swift source file will look like a
+    // closure due to the curly braces. The internals might be a syntactic
+    // disaster, but we don't actually care.
+    if let closure = syntax.as(ClosureExprSyntax.self) {
+      if let file = closure.parent?.parent?.parent?.as(SourceFileSyntax.self),
+        let unexpected = file.unexpectedBetweenStatementsAndEndOfFileToken
+      {
+        return .tail(closure, unexpected)
+      }
+
+      return .closure(closure)
+    }
+
+    // We found a string literal; its contents might be JSON.
+    if let stringLit = syntax.as(StringLiteralExprSyntax.self) {
+      guard let text = stringLit.representedLiteralValue else {
+        return nil
+      }
+
+      return .stringLiteral(stringLit, text)
+    }
+
+    // Look further up the syntax tree.
+    if let parent = syntax.parent {
+      return preflightRefactoring(parent)
+    }
+
+    return nil
+  }
+}
+
+extension ConvertJSONToCodableStruct: SyntaxRefactoringCodeActionProvider {
+  static var title = "Create Codable structs from JSON"
+}
+
+extension ConvertJSONToCodableStruct {
+  fileprivate static let indentation = 4
+
+  private static func build(from jsonDictionary: Dictionary<String, Any>) -> DeclSyntax {
```

Comment:
```suggestion
  private static func build(from jsonDictionary: [String: Any]) -> DeclSyntax {
```
 
Comment for `Dictionary<String, Any>` applies to a few more places in the file as well.

---

Also: Do we need this function at all? Doesn’t seem like it’s doing a whole lot and I think it’s only called once.

Reply:
Nah, we can remove it.